st2common/st2common/runners/utils.py

```diff
@@ -32,23 +32,34 @@
 
 LOG = logging.getLogger(__name__)
 
+# Maps logger name to the actual logger instance
+# We re-use loggers for the same actions to make sure only a single instance exists for a
+# particular action. This way we avoid duplicate log messages, etc.
+LOGGERS = {}
+
 
 def get_logger_for_python_runner_action(action_name, log_level='debug'):
     """
     Set up a logger which logs all the messages with level DEBUG and above to stderr.
     """
-    level_name = log_level.upper()
-    log_level_constant = getattr(stdlib_logging, level_name, stdlib_logging.DEBUG)
     logger_name = 'actions.python.%s' % (action_name)
-    logger = logging.getLogger(logger_name)
 
-    console = stdlib_logging.StreamHandler()
-    console.setLevel(log_level_constant)
+    if logger_name not in LOGGERS:
+        level_name = log_level.upper()
+        log_level_constant = getattr(stdlib_logging, level_name, stdlib_logging.DEBUG)
+        logger = logging.getLogger(logger_name)
+
+        console = stdlib_logging.StreamHandler()
+        console.setLevel(log_level_constant)
+
+        formatter = stdlib_logging.Formatter('%(name)-12s: %(levelname)-8s %(message)s')
+        console.setFormatter(formatter)
+        logger.addHandler(console)
+        logger.setLevel(log_level_constant)
 
-    formatter = stdlib_logging.Formatter('%(name)-12s: %(levelname)-8s %(message)s')
-    console.setFormatter(formatter)
-    logger.addHandler(console)
-    logger.setLevel(log_level_constant)
+        LOGGERS[logger_name] = logger
+    else:
+        logger = LOGGERS[logger_name]
 
     return logger
```

Comment:
Will the shared logger work fine when running two actions of same kind simultaneously (mistral parallel for example)?

Reply:
As described in the PR - it's per action process wrapper basis (aka per action invocation basis) so it's not an issue (two actions = two processes).

Even if it was shared among all the actions with the same name it also wouldn't be an issue. It would potentially only become an issue if it was per action runner process / service basis and there were many different actions (cache could grow kinda large in that scenario).